### ISSUE TITLE
Creating a master MOOSE application on an MPI subcomunicator does not work

### ISSUE THREAD

**Issue by permcody** [idaholab/moose#6221]
Here is a fixed code snippet submitted by one of our users that does not work when I believe that it should.

The code makes it through the `MooseInit` call but hangs once `app->run()` is called. My belief is that we have at least one place in MOOSE that assumes `MPI_COMM_WORLD`. The second theory is that there is a problem either in libMesh or PETSc somewhere along the same lines. libMesh should be able to run on a subcommunicator and has logic to handle the case where MPI is already initialized elsewhere (which it would be in any subcommunicator run). This isn't a high priority for us, but should be fixed at some point.

```
int main(int argc, char *argv[])
{
  MPI_Comm   Sub_Comm;       /* intra-communicator of local sub-group */
  int ColorKey;
  int rank;
  int sub_rank;

  //Initialize MPI;
  MPI_Init(&argc, &argv);
  MPI_Comm_rank(MPI_COMM_WORLD, &rank);

  /* User code must generate membershipKey in the range [0, 1] */
  ColorKey = rank % 2;

  /* Build intra-communicator for local sub-group */
  MPI_Comm_split(MPI_COMM_WORLD, ColorKey, rank, &Sub_Comm);
  MPI_Comm_rank(Sub_Comm, &sub_rank);

  if (ColorKey==0)
  {
    std::cout<<"rank: "<<rank<<std::endl;
    std::cout<<"sub_rank: "<<sub_rank<<std::endl;
    std::cout<<"ColorKey: "<<ColorKey<<std::endl;

    // Initialize MPI, solvers and MOOSE
    MooseInit init(argc, argv, Sub_Comm);

    std::cout << "After MooseInit" << std::endl;

    // Register this application's MooseApp and any it depends on
    MooseTestApp::registerApps();

    // This creates dynamic memory that we're responsible for deleting
    MooseApp * app = AppFactory::createApp("MooseTestApp", argc, argv);

    app->setCheckUnusedFlag(true);
    app->setErrorOverridden();

    // Execute the application
    app->run();

    std::cout << "After Run" << std::endl;

    // Free up the memory we created earlier
    delete app;
  }

  MPI_Finalize();
}
```

**Comment by permcody** [comment]
@rppawlo - A user reported this a few weeks back and I was unable to get a very simple test code like this working. If you look at it, it's just creating MOOSE with a subcommunicator instead of MPI_COMM_WORLD like we usually do. It seems to make it through initialization ok but hangs when we call run. Aren't you doing something similar though in VERA and it's working for you? Just curious.

**Comment by rppawlo** [comment]
Hi @permcody - we are running MOOSE (bison multiapp) on a subcommunicator and each multiapp is then run on a subset of that subcomm.  I was seeing a similar seg fault in December when we tried to update Bison/MOOSE until @aeslaughter put in the fake top level app for us. Then this issue went away.  I'm pretty sure this is because we don't solve a top level app and Andrew's work put in the "no-solve" option.  We just solve the underlying apps using the MultiApp object directly.

I'm pretty sure I know what the  problem is.  The createApp() method uses mpi comm world (or at least it did last time I checked). To get around this, here's what we do:
```
       Teuchos::RCP<MooseInit> bison_init = 
        Teuchos::rcp(new MooseInit(argc, argv, (*Teuchos::rcp_dynamic_cast<const Teuchos::MpiComm<int> >(bisonSubcomm)->getRawMpiComm())()));
      delete [] bison_argv;

      // Register this application's MooseApp and any it depends on
      CASL_MOOSEApp::registerApps();

      // used to call createApp here, but now had to switch to
      // create() to set the correct subcomm for the coupled problem.
      // Pulled the code from createApp but switched out the comm.
      Teuchos::RCP<MooseApp> app;
      {
        InputParameters app_params = AppFactory::instance().getValidParams("CASL_MOOSEApp");
        std::shared_ptr<CommandLine> command_line(new CommandLine(argc, argv));
        app_params.set<std::shared_ptr<CommandLine> >("_command_line") = command_line;
        app = Teuchos::rcp(AppFactory::instance().create("CASL_MOOSEApp","main",app_params,
              (*Teuchos::rcp_dynamic_cast<const Teuchos::MpiComm<int> >(bisonSubcomm)->getRawMpiComm())()));
      }

      // These two lines create a fake top level app
      // See idaholab/moose github item #6126
      app->parameters().set<bool>("minimal") = true;
      app->run();

      Factory & factory = app->getFactory();
      InputParameters params = factory.getValidParams("TransientMultiApp");
      params.set<FEProblem *>("_fe_problem") = &(app->getExecutioner()->feProblem());
      params.set<SubProblem*>("_subproblem")  = &(app->getExecutioner()->feProblem());
      params.set<MooseEnum>("app_type") = "CASL_MOOSEApp";
      params.set<MPI_Comm>("_mpi_comm") = (*Teuchos::rcp_dynamic_cast<const Teuchos::MpiComm<int> >(bisonSubcomm)->getRawMpiComm())();
      params.set<bool>("output_in_position") = true;
      params.set<THREAD_ID>("_tid") = 0;
      if (inline_mode_) {
	params.set<bool>("sub_cycling") = true;
	params.set<bool>("print_sub_cycles") = true;
      }
.
.
.
      // Build the multiApp
      {
        std::shared_ptr<MooseObject> mo = factory.create("TransientMultiApp", "Tiamat_MultiApp", params);
        std::shared_ptr<TransientMultiApp> tma = std::dynamic_pointer_cast<TransientMultiApp>(mo);
        multiApp_ = Teuchos::rcp(tma);
      }
```
Our app->run() is a no-op at the top level so this may be keeping us from hanging. I call it to make sure things are initialized, but you may not need to.  

The key point is that we call create() with the correct mpi subcomm. At some point someone removed the comm parameter from createApp() and forced comm world on everyone. Here's the code from AppFactory:
```
MooseApp *
AppFactory::createApp(std::string app_type, int argc, char ** argv)
{
  MooseSharedPointer<CommandLine> command_line(new CommandLine(argc, argv));
  InputParameters app_params = AppFactory::instance().getValidParams(app_type);

  app_params.set<int>("_argc") = argc;
  app_params.set<char**>("_argv") = argv;
  app_params.set<MooseSharedPointer<CommandLine> >("_command_line") = command_line;

  MooseApp * app = AppFactory::instance().create(app_type, "main", app_params, MPI_COMM_WORLD);
  return app;
}
```

So they may be hanging when createApp changes the comm. It would be great if someone could restore the createApp() function to allow an mpi communicator to be passed in as a parameter (see comment in code snipet above). Again, after this we can directly run the MultiApps on their own MPI process subsets without issue directly from the MultiApp interface.

Hope this helps.

**Comment by permcody** [comment]
Thanks - that's a lot of help. When we build the test app(/example/tutorial/unit test). We'll make sure this is all cleaned up. It's easy to have these types of regressions when we don't have tests.

**Comment by friedmud** [comment]
TLDR: Use AppFactory::create()

There are two ways to create a MooseApp...

1.  AppFactory::createApp(): This is only supposed to be called by main()
for our normal MOOSE-based applications.  It's really only there as a
helper to simplify what goes in main().  If you need to do something more
complicated than just creating the main MooseApp with COMM_WORLD then you
_shouldn't_ call this function.

2.  AppFactory::create():
https://github.com/idaholab/moose/blob/devel/framework/src/base/AppFactory.C#L54
This is a slightly lower level function that is called by createApp() that
allows you to specify the MPI_COMM.  You'll need to construct your Params
object yourself (similar to what's in AppFactory::createApp() ).  This is
what you guys need to call.  You can see how it's done here:
https://github.com/idaholab/moose/blob/devel/framework/src/multiapps/MultiApp.C#L455

Derek

**Comment by permcody** [comment]
I have a small patch to MOOSE that actually makes this work. While Derek's comment is close, it's not quite right in that the AppFactory::create() is not static. So you are still forced to create the main application with MPI_COMM_WORLD. My patch will allow you to create the main on a sub communicator.
